### ISSUE TITLE
opensc-pkcs11.so doesn't get installed via apt-get

### ISSUE THREAD

**Issue by victoria-tyfone** [OpenSC/OpenSC#470]
I've decided to 'apt-get install opensc' and all the tools work great except that I can't add my device into "Secure Devices" of Firefox Browser because opensc-pkcs11.so isn't in /usr/lib.  Searched my entire system and don't see it anywhere.

Any plans to match the files that come across via 'apt-get install opensc' to the files that get installed via 'make install'?

I'm running the latest Raspbian image on RPi 2.

Thanks!

**Comment by dengert** [comment]
On 5/20/2015 6:26 PM, victoria-tyfone wrote:
> I've decided to 'apt-get install opensc' and all the tools work great except that I can't add my device into "Secure Devices" of Firefox Browser because opensc-pkcs11.so isn't in /usr/lib. Searched my
> entire system and don't see it anywhere.
>
> Any plans to match the files that come across via 'apt-get install opensc' to the files that get installed via 'make install'?

>
> I'm running the latest Raspbian image on RPi 2.

Sounds like a Raspbian packaging issue.
Try:
apt-get install opensc-pkcs11


   http://mirrordirectortest.raspbian.org/raspbian/pool/main/o/opensc/

   http://mirrordirectortest.raspbian.org/raspbian/pool/main/o/opensc/opensc_0.14.0-2_armhf.deb

does not have opensc-pkcs11 and separate package is:

   http://mirrordirectortest.raspbian.org/raspbian/pool/main/o/opensc/opensc-pkcs11_0.14.0-2_armhf.deb






>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/470>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by victoria-tyfone** [comment]
Actually, I redirected apt-get to install from sid (https://packages.debian.org/sid/utils/opensc) because I need the latest OpenSC (0.15.0) and other more recent packages.  And this install already brings in opensc-pkcs11, since it's marked as a dependency.
I tried to install again to confirm and apt-get does tell me that everything has already been installed.

So, your expectation is that opensc-pkcs11.so should have appeared in /usr/lib?

If so, where might it have gone?

Thanks!

**Comment by dengert** [comment]
On 5/20/2015 11:46 PM, victoria-tyfone wrote:
> Actually, I redirected apt-get to install from sid (https://packages.debian.org/sid/utils/opensc) because I need the latest OpenSC (0.15.0) and other more recent packages. And this install already
> brings in opensc-pkcs11, since it's marked as a dependency.
> I tried to install again to confirm and apt-get does tell me that everything has already been installed.
>
> So, your expectation is that opensc-pkcs11.so should have appeared in /usr/lib?
>
> If so, where might it have gone?

The Raspberry Pi use the arm processor...

https://packages.debian.org/sid/opensc-pkcs11

https://packages.debian.org/sid/arm64/opensc-pkcs11/filelist

says it should be:
/usr/lib/aarch64-linux-gnu/pkcs11/opensc-pkcs11.so




>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/470#issuecomment-104135143>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
[updated answer]

On 5/20/2015 11:46 PM, victoria-tyfone wrote:
> Actually, I redirected apt-get to install from sid (https://packages.debian.org/sid/utils/opensc) because I need the latest OpenSC (0.15.0) and other more recent packages. And this install already
> brings in opensc-pkcs11, since it's marked as a dependency.
> I tried to install again to confirm and apt-get does tell me that everything has already been installed.
>
> So, your expectation is that opensc-pkcs11.so should have appeared in /usr/lib?
>
> If so, where might it have gone?

The Raspberry Pi use the arm processor...

https://packages.debian.org/sid/opensc-pkcs11

https://packages.debian.org/sid/arm64/opensc-pkcs11/filelist

says it should be:
/usr/lib/aarch64-linux-gnu/pkcs11/opensc-pkcs11.so


If this is the RPI2, then it should be:

https://packages.debian.org/sid/armhf/opensc-pkcs11/filelist

/usr/lib/arm-linux-gnueabihf/pkcs11/opensc-pkcs11.so


> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/470#issuecomment-104135143>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
On 5/20/2015 11:46 PM, victoria-tyfone wrote:
> Actually, I redirected apt-get to install from sid (https://packages.debian.org/sid/utils/opensc) because I need the latest OpenSC (0.15.0) and other more recent packages. And this install already
> brings in opensc-pkcs11, since it's marked as a dependency.
> I tried to install again to confirm and apt-get does tell me that everything has already been installed.

In you first note, you indicated your were running  Raspbian image on RPi 2.

Reading these:

  https://wiki.debian.org/RaspberryPi

  https://wiki.debian.org/RaspberryPi2

implies that the Raspbian releases may differ from the Debian release, so just switching to debian sid may
not be the correct way to go, but the RPi2 may not have as many problems as the older processors. (I have an older version.)

You may want to download the debian source for OpenSC and try building.
Or just build OpenSC into /usr/local



>
> So, your expectation is that opensc-pkcs11.so should have appeared in /usr/lib?
>
> If so, where might it have gone?
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/470#issuecomment-104135143>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by victoria-tyfone** [comment]
Thank you very much for your time, Doug!

I reinstalled opensc and opensc-pkcs11.so is in /usr/lib/arm-linux-gnueabihf/pkcs11 this time around.  Additionally, Iceweasel loads it just fine.  Not sure why it wasn't showing up when I searched /usr before.

I did build and install OpenSC on RPi 2 previously as well, hence my initial confusion, because 'make install' does copy opensc-pkcs11.so into /usr/lib.  But it does not, for example, copy pkcs11-spy.so, that lib stays in src/pkcs11/.libs.

What is the reasoning behind this inconsistency in the installs?

Thanks!

**Comment by dengert** [comment]
On 5/22/2015 5:25 PM, victoria-tyfone wrote:
> Thank you very much for your time, Doug!
>
> I reinstalled opensc and opensc-pkcs11.so is in /usr/lib/arm-linux-gnueabihf/pkcs11 this time around. Additionally, Iceweasel loads it just fine. Not sure why it wasn't showing up when I searched /usr
> before.
>
> I did build and install OpenSC on RPi 2 previously as well, hence my initial confusion, because 'make install' does copy opensc-pkcs11.so into /usr/lib. But it does not, for example, copy
> pkcs11-spy.so, that lib stays in src/pkcs11/.libs.

I have just finished updating my Raspberry Pi type B,
Linux raspberrypi 3.12.28+ #709 PREEMPT Mon Sep 8 15:28:00 BST 2014 armv6l GNU/Linux

and built the Debian SID version, with these instructions (cd and mkdir not shown)

add to /etc/apt/sources.list:
deb-src http://ftp.us.debian.org/debian/ sid main

sudo apt-get  source opensc
dpkg-buildpackage -rfakeroot -uc -b

sudo dpkg -i opensc-pkcs11_0.15.0-1_armhf.deb
sudo dpkg -i opensc_0.15.0-1_armhf.deb

I had to start with source, the Debian sid binaries get an illegal instruction because Raspbian and Debian produce
different instructions for armhf They did not agree on what processors are armhf.


I see in:

  $ cd /usr/lib/arm-linux-gnueabihf/
  $ ls -lt
total 207820
drwxr-xr-x 2 root root     4096 May 23 01:18 pkcs11
-rw-r--r-- 1 root root  1464032 May 23 01:04 libopensc.so.3.0.0
-rw-r--r-- 1 root root   153272 May 23 01:04 onepin-opensc-pkcs11.so
-rw-r--r-- 1 root root    63028 May 23 01:04 libsmm-local.so.3.0.0
-rw-r--r-- 1 root root   153272 May 23 01:04 opensc-pkcs11.so
-rw-r--r-- 1 root root    55924 May 23 01:04 pkcs11-spy.so
lrwxrwxrwx 1 root root       18 May 23 01:04 libopensc.so.3 -> libopensc.so.3.0.0
lrwxrwxrwx 1 root root       21 May 23 01:04 libsmm-local.so.3 -> libsmm-local.so.3.0.0
[...]
  $ cd pkcs11
  $ ls -l
total 0
lrwxrwxrwx 1 root root 26 May 23 01:04 onepin-opensc-pkcs11.so -> ../onepin-opensc-pkcs11.so
lrwxrwxrwx 1 root root 19 May 23 01:04 opensc-pkcs11.so -> ../opensc-pkcs11.so
lrwxrwxrwx 1 root root 16 May 23 01:04 pkcs11-spy.so -> ../pkcs11-spy.so


pkcs11-spy.so is there.


>
> What is the reasoning behind this inconsistency in the installs?
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/470#issuecomment-104788766>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>
